### PR TITLE
Update the issuer metadata for signed metadata

### PR DESCRIPTION
**Pull request by Ogenbertrand**
Description

Update metadata retrieval to support signed metadata (application/jwt) alongside unsigned (application/json), including sending an Accept header and validating JWS headers (alg, typ: openidvci-issuer-metadata+jwt) and JWT claims (iss, sub, iat, exp).

See: https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0-16.html#name-signed-metadata

closes: https://github.com/keycloak/keycloak/issues/41588

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 1]
```diff
@@ -81,8 +98,17 @@ public Object getConfig() {
...
                 .setAuthorizationServers(List.of(getIssuer(context)))
                 .setCredentialResponseEncryption(getCredentialResponseEncryption(keycloakSession))
-                .setBatchCredentialIssuance(getBatchCredentialIssuance(keycloakSession))
-                .setSignedMetadata(getSignedMetadata(keycloakSession));
+                .setBatchCredentialIssuance(getBatchCredentialIssuance(keycloakSession));
+
+        boolean signedMetadataEnabled = Boolean.parseBoolean(realm.getAttribute(SIGNED_METADATA_ENABLED_ATTR));
+        if (signedMetadataEnabled) {
+            try {
+                return generateSignedMetadata(issuer, keycloakSession);
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: When signed metadata is returned, the response should include `Content-Type: application/jwt` header, but the current implementation doesn't set this header.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 2]
```diff
@@ -81,8 +98,17 @@ public Object getConfig() {
...
+            } catch (Exception e) {
+                LOGGER.errorf("Failed to generate signed metadata for realm: %s", realm.getName(), e);
+                throw new RuntimeException("Unable to generate signed metadata", e);
+            }
+        }
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The error handling catches all exceptions and throws a generic RuntimeException. Consider using more specific exception types or providing more detailed error information for debugging.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 3]
```diff
@@ -104,9 +130,92 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+                .stream()
+                .filter(alg -> !alg.startsWith("HS"))
+                .collect(Collectors.toList());
+        if (supportedAlgorithms.isEmpty()) {
+            throw new IllegalStateException("No asymmetric signing algorithms available for realm: " + realm.getName());
+        }
+
+        String alg = Optional.ofNullable(realm.getAttribute(SIGNED_METADATA_ALG_ATTR))
+                .filter(supportedAlgorithms::contains)
+                .orElse(supportedAlgorithms.get(0));
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The algorithm selection logic is good, but consider adding validation to ensure the selected algorithm is appropriate for the key type (e.g., RSA keys should use RSA algorithms).

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 4]
```diff
@@ -104,9 +130,92 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+                .kid(keyWrapper.getKid());
+
+        // Add x5c if a certificate or certificate chain is available
+        if (keyWrapper.getCertificateChain() != null && !keyWrapper.getCertificateChain().isEmpty()) {
+            jwsBuilder.x5c(keyWrapper.getCertificateChain());
+        } else if (keyWrapper.getCertificate() != null) {
+            jwsBuilder.x5c(List.of(keyWrapper.getCertificate()));
+        } else {
+            LOGGER.warnf("No certificate or certificate chain available for x5c header in realm: %s", realm.getName());
+        }
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The certificate handling is well-implemented, but consider adding validation to ensure the certificate chain is valid and properly ordered.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 5]
```diff
@@ -104,9 +130,92 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+    /**
+     * Generates signed metadata as a JWS.
+     *
+     * @param metadata The CredentialIssuer metadata object to sign.
+     * @param session  The Keycloak session.
+     * @return The compact JWS string.
+     * @throws IllegalStateException if generation fails due to configuration or signing issues.
+     */
+    public String generateSignedMetadata(CredentialIssuer metadata, KeycloakSession session) {
+
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The `generateSignedMetadata` method is quite long. Consider breaking it into smaller, more focused methods for better readability and testability.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java
Comment (Awambeng) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java:1]: The test implementation is comprehensive and covers the main scenarios, but consider adding tests for:
- Edge cases with invalid configuration

**Review by Awambeng — CHANGES_REQUESTED** [review]
The `well-known` endpoint currently only produces `application/json` and doesn't check the Accept header to determine whether to return signed or unsigned metadata. According to the spec, the server should check the Accept header and return `application/jwt` when requested.

**Review by forkimenjeckayang — CHANGES_REQUESTED** [review]
Just a few remarks to leave, as i started working on this without the knowledge of you already working on it.

### The implementation does not support content negotiation as required by OID4VCI specification section 11.2.2.

Current implementation only supports realm-level configuration, not HTTP-level content negotiation.

**Missing Requirements**:
- No Accept header processing for application/json vs application/jwt
- No proper Content-Type header handling
- No support for signed metadata via Accept: application/jwt

**Specification Quote**: "The Credential Issuer MUST respond with HTTP Status Code 200 and return the Credential Issuer Metadata containing the parameters defined in Section 11.2.4 as either an unsigned JSON document using the media type application/json, or a signed JSON Web Token (JWT) containing the Credential Issuer Metadata in its payload using the media type application/jwt."

### No support for `Accept-Language` header processing as required by the specification.

Current implementation has no internationalization support whatsoever.

**Missing Requirements**:
- No Accept-Language header processing
- No Content-Language response header
- No internationalized display data support

**Specification Quote**: "The Wallet is RECOMMENDED to send an Accept-Language header in the HTTP GET request to indicate the language(s) preferred for display."

**Comment by Ogenbertrand** [comment]
> Just a few remarks to leave, as i started working on this without the knowledge of you already working on it.
> 
> ### The implementation does not support content negotiation as required by OID4VCI specification section 11.2.2.
> Current implementation only supports realm-level configuration, not HTTP-level content negotiation.
> 
> **Missing Requirements**:
> 
> * No Accept header processing for application/json vs application/jwt
> * No proper Content-Type header handling
> * No support for signed metadata via Accept: application/jwt
> 
> **Specification Quote**: "The Credential Issuer MUST respond with HTTP Status Code 200 and return the Credential Issuer Metadata containing the parameters defined in Section 11.2.4 as either an unsigned JSON document using the media type application/json, or a signed JSON Web Token (JWT) containing the Credential Issuer Metadata in its payload using the media type application/jwt."
> 
> ### No support for `Accept-Language` header processing as required by the specification.
> Current implementation has no internationalization support whatsoever.
> 
> **Missing Requirements**:
> 
> * No Accept-Language header processing
> * No Content-Language response header
> * No internationalized display data support
> 
> **Specification Quote**: "The Wallet is RECOMMENDED to send an Accept-Language header in the HTTP GET request to indicate the language(s) preferred for display."

Thank you for the comments Dear @forkimenjeckayang !
I confrim that my logic is currently missing the Content Type header when making a request which i am already handling now. but for the Accept-Language Keyclaok don't have any logic for that reasons i don't know. it's not part of the change for this task.

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java
Comment (forkimenjeckayang) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java:1]: Still no implementation of `Accept-Language` header processing.
I'll still insist we need this. Keep in mind conformance tests will be performed with different issuers possibly. And also, this will detrmine what the `display` object will look like in the credentilal issuer metadata.

On the other hand
It is up to the Credential Issuer whether to:
- send a subset the metadata containing internationalized display data for one or all of the requested languages and indicate returned languages using the HTTP Content-Language Header, or
- ignore the Accept-Language Header and send all supported languages or any chosen subset.

So if you decide to proceed:
This requires:
1. Processing Accept-Language headers in the HTTP request
2. Returning Content-Language response headers
3. Filtering display information based on language preferences
4. Integration with realm internationalization settings

If not then it's fine. 
@Ogenbertrand WDYT?

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java [hunk 6]
```diff
@@ -222,28 +226,61 @@ public Response getVersionPreflight(final @PathParam("realm") String name,
...
+            try {
+                signedJwt = oid4vcProvider.generateSignedMetadata((CredentialIssuer) config, session);
+            } catch (Exception e) {
+                logger.warnf(e, "Failed to generate signed metadata for realm: %s", realm.getName());
+            }
+            if (signedJwt != null) {
+                responseBuilder = Response.ok(signedJwt).type(org.keycloak.utils.MediaType.APPLICATION_JWT);
+            } else {
+                logger.debugf("Falling back to JSON response due to signed metadata failure for realm: %s", realm.getName());
+                responseBuilder = Response.ok(config).type(MediaType.APPLICATION_JSON);
```
Comment (forkimenjeckayang) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: Incorrect error handling for content negotiation. When a client requests application/jwt but signing fails, the implementation should return 406 Not Acceptable, not fall back to JSON.

The specification states: `It is RECOMMENDED for Credential Issuers to respond with a Content-Type matching to the Wallet's requested Accept header when the requested content type is supported.`

Current implementation:
- Client requests: Accept: application/jwt
- Server fails to sign → Returns application/json 

Should be:
- Client requests: Accept: application/jwt  
- Server fails to sign → Returns 406 Not Acceptable

The fallback to JSON violates the specification's content negotiation requirements.

**Review by forkimenjeckayang — CHANGES_REQUESTED** [review]
Left a few comments @Ogenbertrand 
Please check

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java
Comment (Ogenbertrand) [reply services/src/main/java/org/keycloak/services/resources/RealmsResource.java:1]: Dear @forkimenjeckayang reasons why i mentioned no need for **Accept-Language** is cause if you check the past versions of the specification like even going backward to 13, 14 and 15 the **Accept-Language** has always been there and KC hasn't considered it. I can't really explain why the don't maybe @IngridPuppet could help explain the reason why. That said the conformance test has always worked even without the **Accept-Language** logic. So i'll suggest we keep things the way they are now.

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java (hunk 6 above)
Comment (Ogenbertrand) [reply services/src/main/java/org/keycloak/services/resources/RealmsResource.java:275]: Yes you are right !
fixed.

Comment (Ogenbertrand) [reply services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: The specification doesn't explicitly mandate behavior for signed metadata failures, but it does state:

"The Credential Issuer MUST support returning metadata in an unsigned form 'application/json'"

"MAY support returning it in a signed form 'application/jwt'"

If signed metadata generation fails, the issuer should either:

Retrun an Error (if signed metadata is required by policy)

Fall back to unsigned JSON but return it with application/json content type

If the client strictly accepts only application/jwt (without listing application/json as an alternative), it may reject the response, but the issuer's responsibility is fulfilled by providing the required unsigned fallback.

Comment (forkimenjeckayang) [reply services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: @Ogenbertrand , I get your point. Just wanted us to treat it as strict given that It's a RECOMMENDATION. A client shouldn't ask for a specific format then we send another. The spec does not mandate it so , if you wish to stay with the fallback , that's okay.

**Review by forkimenjeckayang — APPROVED** [review]
My concerns have been addressed

**Review by Awambeng — APPROVED** [review]

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 7]
```diff
@@ -53,6 +61,12 @@
  */
 public class OID4VCIssuerWellKnownProvider implements WellKnownProvider {
 
+    // Realm attributes for signed metadata configuration
+    public static final String SIGNED_METADATA_ENABLED_ATTR = "oid4vci.signed_metadata.enabled";
+    public static final String SIGNED_METADATA_EXPIRATION_ATTR = "oid4vci.signed_metadata.exp";
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: Rename this to lifespan instead of just "expiration" for clarity.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 8]
```diff
@@ -53,6 +61,12 @@
...
+    public static final String SIGNED_METADATA_ISS_ATTR = "oid4vci.signed_metadata.iss";
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The issuer field really does not need to be configurable.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 9]
```diff
@@ -104,13 +118,126 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+                .orElse(supportedAlgorithms.get(0));
     }
 
+    private JsonWebToken createMetadataJwt(CredentialIssuer metadata, RealmModel realm) {
+        JsonWebToken jwt = new JsonWebToken();
+
+        // Set standard JWT claims
+        jwt.subject(metadata.getCredentialIssuer());
+        jwt.iat((long) Time.currentTime());
+        jwt.issuedNow();
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: Do you need to do both?

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 10]
```diff
@@ -104,13 +118,126 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to sign metadata", e);
+        }
+    }
+
+    private String getSigningAlgorithm(RealmModel realm, KeycloakSession session) {
+        List<String> supportedAlgorithms = getSupportedSignatureAlgorithms(session)
+                .stream()
+                .filter(alg -> !alg.startsWith("HS")) // Filter out symmetric algorithms
+                .collect(Collectors.toList());
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: This "filter out" is not really reliable. Create a static list of asymetric algorithms that you intersect with the supported algorithms in the session.

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 11]
```diff
@@ -104,13 +118,126 @@ private CredentialIssuer.BatchCredentialIssuance getBatchCredentialIssuance(Keyc
...
+            throw new IllegalStateException("No asymmetric signing algorithms available for realm: " + realm.getName());
+        }
+
+        String configuredAlg = realm.getAttribute(SIGNED_METADATA_ALG_ATTR);
+        if (configuredAlg != null && !supportedAlgorithms.contains(configuredAlg)) {
+            throw new IllegalStateException("Configured signing algorithm '" + configuredAlg + "' is not supported for realm: " + realm.getName());
+        }
+
+        return Optional.ofNullable(configuredAlg)
+                .orElse(supportedAlgorithms.get(0));
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: It is preferable to use `Algorithm.RS256`. I ran into a situation where this first item is `HS512`, which is unwanted.

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java [hunk 12]
```diff
@@ -222,28 +226,61 @@ public Response getVersionPreflight(final @PathParam("realm") String name,
 
     @GET
     @Path("{realm}/.well-known/{alias}")
-    @Produces(MediaType.APPLICATION_JSON)
+    @Produces({MediaType.APPLICATION_JSON, org.keycloak.utils.MediaType.APPLICATION_JWT})
+    @NoCache
     public Response getWellKnown(final @PathParam("realm") String name,
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: Do you need this NoCache?

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java [hunk 13]
```diff
@@ -222,28 +226,61 @@ public Response getVersionPreflight(final @PathParam("realm") String name,
...
+            // Generate signed JWT metadata
+            OID4VCIssuerWellKnownProvider oid4vcProvider = (OID4VCIssuerWellKnownProvider) wellKnown;
+            String signedJwt = null;
+            try {
+                signedJwt = oid4vcProvider.generateSignedMetadata((CredentialIssuer) config, session);
+            } catch (Exception e) {
+                logger.warnf(e, "Failed to generate signed metadata for realm: %s", realm.getName());
+            }
+            if (signedJwt != null) {
+                responseBuilder = Response.ok(signedJwt).type(org.keycloak.utils.MediaType.APPLICATION_JWT);
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: No OID4VCI-specific code can live in this class. `OID4VCIssuerWellKnownProvider` has access to the request headers via the session object and may make this reasoning on its own.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java [hunk 14]
```diff
@@ -101,6 +126,188 @@ public void configureTestRealm(RealmRepresentation testRealm) {
...
+        HttpGet getJsonMetadata = new HttpGet(wellKnownUri);
+        getJsonMetadata.addHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON);
+        try (CloseableHttpResponse response = httpClient.execute(getJsonMetadata)) {
+            assertEquals(HttpStatus.SC_OK, response.getStatusLine().getStatusCode());
+            assertEquals("Content-Type should be application/json", MediaType.APPLICATION_JSON,
+                    response.getFirstHeader(HttpHeaders.CONTENT_TYPE).getValue());
+            String json = EntityUtils.toString(response.getEntity(), StandardCharsets.UTF_8);
+            LOGGER.infof("Response should be a JSON string: %s", json);
+            CredentialIssuer issuer = JsonSerialization.readValue(json, CredentialIssuer.class);
+            LOGGER.infof("Response should be a CredentialIssuer object: %s", json);
```
Comment (IngridPuppet) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java]: Logging in tests is not really necessary.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java [hunk 15]
```diff
@@ -101,6 +126,188 @@ public void configureTestRealm(RealmRepresentation testRealm) {
...
+        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
+        String wellKnownUri = OAuthClient.AUTH_SERVER_ROOT + "/realms/" + realmName + "/.well-known/openid-credential-issuer";
+
+        // Test Case 1: Unsigned metadata with application/json (mandatory per spec)
+        testingClient.server(realmName).run(session -> {
+            RealmModel realm = session.getContext().getRealm();
+            realm.setAttribute(OID4VCIssuerWellKnownProvider.SIGNED_METADATA_ENABLED_ATTR, "true");
+            realm.setAttribute(OID4VCIssuerWellKnownProvider.SIGNED_METADATA_ALG_ATTR, "RS256");
+            realm.setAttribute(OID4VCIssuerWellKnownProvider.SIGNED_METADATA_EXPIRATION_ATTR, "3600");
+            realm.setAttribute(OID4VCIssuerWellKnownProvider.SIGNED_METADATA_ISS_ATTR, "https://example.com/issuer");
```
Comment (IngridPuppet) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java]: I'm not sure this is properly formatted.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java [hunk 16]
```diff
@@ -101,6 +126,188 @@ public void configureTestRealm(RealmRepresentation testRealm) {
...
 
+    @Test
+    public void testSignedMetadataGeneration() {
+        runSignedMetadataTest(testingClient, TEST_REALM_NAME, getRealmPath(TEST_REALM_NAME));
+    }
+
+    private static void runSignedMetadataTest(KeycloakTestingClient testingClient, String realmName, String expectedIssuer) {
+        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
+        String wellKnownUri = OAuthClient.AUTH_SERVER_ROOT + "/realms/" + realmName + "/.well-known/openid-credential-issuer";
+
```
Comment (IngridPuppet) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java]: Create multiple tests for each test case.

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java (hunk 8 above)
Comment (Ogenbertrand) [reply services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: Just to be clare, are you refering to all the three constants or just the SIGNED_METADATA_ISS_ATTR?

Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: Just `SIGNED_METADATA_ISS_ATTR`.

Comment (Ogenbertrand) [reply services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: Alright then it done @IngridPuppet !!!

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java [hunk 17]
```diff
@@ -248,39 +243,13 @@ public Response getWellKnown(final @PathParam("realm") String name,
...
-                responseBuilder = Response.ok(signedJwt).type(org.keycloak.utils.MediaType.APPLICATION_JWT);
-            } else {
-                logger.debugf("Falling back to JSON response due to signed metadata failure for realm: %s", realm.getName());
-                responseBuilder = Response.ok(config).type(MediaType.APPLICATION_JSON);
-            }
-        } else {
-            // Default to JSON response
-            responseBuilder = Response.ok(config).type(MediaType.APPLICATION_JSON);
+        if (wellKnown instanceof OID4VCIssuerWellKnownProvider) {
+            return ((OID4VCIssuerWellKnownProvider) wellKnown).getMetadataResponse(session);
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: No OID4VC-specific code should live in this class. `OID4VCIssuerWellKnownProvider` overrides a `getConfig` method where this can be implemented. Ideally, nothing should change in this file, apart from:

```java
@Produces({MediaType.APPLICATION_JSON, org.keycloak.utils.MediaType.APPLICATION_JWT})
```

Remove `org.keycloak.utils.`, it gives ChatGPT away.

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java [hunk 18]
```diff
@@ -108,6 +107,8 @@ public void close() {
     @Override
     public Object getConfig() {
         KeycloakContext context = keycloakSession.getContext();
+        RealmModel realm = keycloakSession.getContext().getRealm();
+        
         CredentialIssuer issuer = new CredentialIssuer()
                 .setCredentialIssuer(getIssuer(context))
                 .setCredentialEndpoint(getCredentialsEndpoint(context))
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java]: The `getConfig` method only returned unsigned metadata in the past, but now also supports returning a signed string, which is fine. However, the `getConfig` method was previously called across the database only to get unsigned metadata. Could you extract this CredentialIssuer object construction into a different method that you call here and replace occurences of calls to `getConfig` across the codebase?

The new method can return a `CredentialIssuer` directly:

```java
public CredentialIssuer getIssuerMetadata() {
```

File: services/src/main/java/org/keycloak/services/resources/RealmsResource.java [hunk 19]
```diff
@@ -247,8 +248,19 @@ public static Response getWellKnownResponse(KeycloakSession session, String name
...
+            Response.ResponseBuilder responseBuilder;
+
+            // Check if the provider returned a JWT string or JSON object
+            if (config instanceof String) {
+                // Provider returned a JWT string
+                responseBuilder = Response.ok(config).type(APPLICATION_JWT);
+            } else {
+                // Provider returned a JSON object
+                responseBuilder = Response.ok(config).type(MediaType.APPLICATION_JSON);
+            }
```
Comment (IngridPuppet) [inline services/src/main/java/org/keycloak/services/resources/RealmsResource.java]: The ternary operator is shorter.

```java
responseBuilder = Response.ok(config).type(config instanceof String ? APPLICATION_JWT : APPLICATION_JSON);
```

**Review by IngridPuppet — APPROVED** [review]
Thank you @Ogenbertrand @forkimenjeckayang.

**Comment by forkimenjeckayang** [comment]
KC PR: https://github.com/keycloak/keycloak/pull/42428
